tests/test_resolver.py

```diff
@@ -1,157 +1,281 @@
 # -*- coding: utf-8 -*-
 
-from os.path import join as pjoin
-from pathlib import Path
-from tempfile import TemporaryDirectory
+import os
+import shutil
 
-from tests.base import TestCase, assets, main, copy_of_directory
+from pathlib import (
+    Path
+)
+from unittest import (
+    mock
+)
+
+import pytest
+
+from tests.base import (
+    assets, 
+    main
+)
 
 from ocrd.resolver import Resolver
-from ocrd_utils import pushd_popd, initLogging
+from ocrd_utils import pushd_popd
+
+# set pylint once on module level
+# pylint: disable=protected-access
 
 METS_HEROLD = assets.url_of('SBB0000F29300010000/data/mets.xml')
 FOLDER_KANT = assets.path_to('kant_aufklaerung_1784')
+DATA_KANT = {'mets.xml': (os.path.join(FOLDER_KANT, 'data', 'mets.xml'), 'text/xml'),
+             'INPUT_0017.tif': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-IMG', 'INPUT_0017.tif'), 'image/tiff'),
+             'INPUT_0020.tif': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-IMG', 'INPUT_0020.tif'), 'image/tiff'),
+             'PAGE_0017_ALTO.xml': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-GT-ALTO', 'PAGE_0017_ALTO.xml'), 'text/xml'),
+             'PAGE_0020_ALTO.xml': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-GT-ALTO', 'PAGE_0020_ALTO.xml'), 'text/xml'),
+             'PAGE_0017_PAGE.xml': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-GT-PAGE', 'PAGE_0017_PAGE.xml'), 'text/xml'),
+             'PAGE_0020_PAGE.xml': (os.path.join(FOLDER_KANT, 'data', 'OCR-D-GT-PAGE', 'PAGE_0020_PAGE.xml'), 'text/xml'),
+             }
+
+
+def _get_kant_data(key):
+    if key in DATA_KANT.keys():
+        (path, mime) = DATA_KANT[key]
+        with open(path, mode='rb') as _file:
+            return (_file.read(), mime)
+
+
+def request_behavior(*args):
+    resp = mock.Mock()
+    resp.status_code = 200
+    resp.headers = {}
+    the_key = args[0].split('/')[-1]
+    if the_key in DATA_KANT:
+        (cnt, mime) = _get_kant_data(the_key)
+        resp.content = cnt
+        resp.headers = {'Content-Type': mime}
+    return resp
+
+
+def test_workspace_from_url_bad():
+    with pytest.raises(Exception) as exc:
+        Resolver().workspace_from_url(None)
+
+    # check exception
+    assert "Must pass 'mets_url'" in str(exc)
+
+
+@mock.patch("requests.get")
+def test_workspace_from_url_kant(mock_request, tmp_path):
```

Comment:
Neat!

Reply:
I still don't quite understand how the `tmp_path` arguments work here, though. Where is the `tempfile.TemporaryDirectory` being set up now?

Reply:
It's a builtin fixture in pytest, cf. https://docs.pytest.org/en/6.2.x/builtin.html. This means that if you use the particular names for fixtures (which can be builtin as in the case of `tmp_path` or looked up in the test suite for matching function names), pytest will notice this and call that fixture for both data and side-effects like setting up/tearing down temporary paths, like here.

Reply:
Got it. _Builtin fixtures_ – or in other words: "Open sesame!"

Reply:
Yup, magic! Or rather clever use of code introspection and generation :-)